docs: add MTEB evaluation guide and update usage.rst (#3477)
UKPLab/sentence-transformers: docs/sentence_transformer/usage/mteb_evaluation.md

# Evaluating SentenceTransformer Models with MTEB

The [Massive Text Embedding Benchmark (MTEB)](https://github.com/embeddings-benchmark/mteb) is a comprehensive benchmark suite for evaluating embedding models across diverse NLP tasks like classification, retrieval, clustering, reranking, and semantic similarity.

> This guide walks you through using MTEB **with SentenceTransformer models for post-training evaluation**. This is *not* designed for use during training loops.

Reviewer: Can you add something like?
```suggestion
This guide walks you through using MTEB **with SentenceTransformer models for post-training evaluation**. This is *not* designed for use during training loops. To fully integrate your model to `MTEB` you can follow this [guide](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_model.md)
```

---

## Installation

Install MTEB and its dependencies:

```bash
pip install mteb
```

> This also installs `sentence-transformers`, `datasets`, and task-specific dependencies.

---

##  Quick Start: One-Line Evaluation

```python
from sentence_transformers import SentenceTransformer
from mteb import MTEB

model = SentenceTransformer("all-MiniLM-L6-v2")

> evaluation = MTEB(tasks=["STSBenchmark"])

Reviewer: Recommended way
```suggestion
tasks = mteb.get_tasks(tasks=["STSBenchmark"])
evaluation = MTEB(tasks=tasks)
```

MTEB supports the following **task families**:

### 1. **Semantic Textual Similarity (STS)**

* **STSBenchmark**: Predict similarity score (0-5) for English sentence pairs.
* **BIOSSES**: Biomedical sentence similarity.
* **SICK-R**: Sentence Involving Compositional Knowledge with human ratings.

### 2. **Classification**

* **AmazonCounterfactualClassification**: Predict product sentiment in counterfactual reviews.
* **TwitterSentiment**: Predict sentiment on tweets (positive/negative).
* **MassiveIntentClassification**: Classify user intent in multi-lingual commands.

### 3. **Retrieval (Information Retrieval)**

* **TREC-COVID**: Retrieve COVID-19 papers for search queries.
* **SciFact**: Retrieve abstracts relevant to scientific claims.
* **NFCorpus**: Retrieve passages for health and nutrition questions.

### 4. **Reranking**

* **MSMARCO**: Rerank candidate passages by relevance to a query.
* **StackExchangeReranking**: Rerank answers on StackExchange questions.

### 5. **Bitext Mining**

* **BUCC**: Identify aligned sentences in multilingual corpora.
* **Tatoeba**: Match translations across 100+ languages.

You can evaluate specific tasks or full categories:

```python

> MTEB(tasks=["SICK-R", "AmazonCounterfactualClassification"])
> MTEB(task_types=["Classification", "Retrieval"])

Reviewer: ```suggestion
mteb.get_tasks(tasks=["SICK-R", "AmazonCounterfactualClassification"])
mteb.get_tasks(task_types=["Classification", "Retrieval"])
```

# Evaluating SentenceTransformer Models with MTEB

The [Massive Text Embedding Benchmark (MTEB)](https://github.com/embeddings-benchmark/mteb) is a comprehensive benchmark suite for evaluating embedding models across diverse NLP tasks like classification, retrieval, clustering, reranking, and semantic similarity.

This guide walks you through using MTEB **with SentenceTransformer models for post-training evaluation**. This is *not* designed for use during training loops. To fully integrate your model to `MTEB` you can follow this [guide](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_model.md)

---

## Installation

Install MTEB and its dependencies:

```bash

> pip install mteb

Reviewer: We're now working on `v2` release that would be breaking
```suggestion
pip install "mteb<2"
```

Reviewer 2: I don't see anything that wouldn't work in v2 as well?

Reviewer: Everything would work, but `mteb.MTEB` will be deprecated

Reviewer 2: It will simply display a deprecation warning (with an upgrade suggestion), which I think is perfectly fine.

Author: so @KennethEnevoldsen  and @Samoed  guys what's your final suggestion shall i change dependencies or no?

Reviewer 2: Keep it as is

```

> This also installs `sentence-transformers`, `datasets`, and task-specific dependencies.

---

##  Quick Start: One-Line Evaluation

```python
from sentence_transformers import SentenceTransformer
from mteb import MTEB

> model = SentenceTransformer("all-MiniLM-L6-v2")

Reviewer: I'm not sure what the best approach is here, because this documentation is for `SentenceTransformer`, not `MTEB`. However, this method might be better since some models could have custom instructions or implementations coming from `MTEB`.

```suggestion
model = mteb.get_model("sentence-transformers/all-MiniLM-L6-v2")
```

Author: ok so i have left it 
at 
`model = SentenceTransformer("all-MiniLM-L6-v2")`

tasks = mteb.get_tasks(tasks=["STSBenchmark"])
evaluation = MTEB(tasks=tasks)
evaluation.run(model, output_folder="results/")
```

> This evaluates your model on the **STS Benchmark**, a Semantic Textual Similarity dataset with human-annotated sentence pairs. Output is saved in `results/`.

Author: @KennethEnevoldsen any particluar task which will be better for this documentation
Just let me know will push the changes

> This evaluates your model on the **STS Benchmark**, a Semantic Textual Similarity dataset with human-annotated sentence pairs. Output is saved in `results/`.

Reviewer 2: STS17 or STS22.v2 are both pretty good. Both are multilingual by default, well-known tasks in the field (SemEval tasks), and are relatively small, so the evaluation won't take too long.

> This evaluates your model on the **STS Benchmark**, a Semantic Textual Similarity dataset with human-annotated sentence pairs. Output is saved in `results/`.

Author: I have taken STS22.v2 @KennethEnevoldsen

---

##  Supported Task Types and Examples

MTEB supports the following **task families**:

### 1. **Semantic Textual Similarity (STS)**

> * **STSBenchmark**: Predict similarity score (0-5) for English sentence pairs.

Reviewer 2: It is not clear that these are just examples - it is not a comprehensive list

Author: I have now added a clear Note to avoid any further confusion

* **BIOSSES**: Biomedical sentence similarity.
* **SICK-R**: Sentence Involving Compositional Knowledge with human ratings.

### 2. **Classification**

* **AmazonCounterfactualClassification**: Predict product sentiment in counterfactual reviews.
* **TwitterSentiment**: Predict sentiment on tweets (positive/negative).
* **MassiveIntentClassification**: Classify user intent in multi-lingual commands.

### 3. **Retrieval (Information Retrieval)**

* **TREC-COVID**: Retrieve COVID-19 papers for search queries.
* **SciFact**: Retrieve abstracts relevant to scientific claims.
* **NFCorpus**: Retrieve passages for health and nutrition questions.

### 4. **Reranking**

* **MSMARCO**: Rerank candidate passages by relevance to a query.
* **StackExchangeReranking**: Rerank answers on StackExchange questions.

### 5. **Bitext Mining**

* **BUCC**: Identify aligned sentences in multilingual corpora.
* **Tatoeba**: Match translations across 100+ languages.

You can evaluate specific tasks or full categories:

```python
mteb.get_tasks(tasks=["SICK-R", "AmazonCounterfactualClassification"])

> mteb.get_tasks(task_types=["Classification", "Retrieval"])

Reviewer 2: This would often get you more than a 100 tasks. I would never recommend it. It is however a good way to find if there are tasks that matche your use case - e.g. if we want to check if there is any medical retrieval tasks for english: 

```
# this needs to be validated:
tasks = mteb.get_tasks(task_types=["Retrieval"], domains = ["Medical"], languages=["eng"])
```

```

---

##  Customize Output and Results Handling

To avoid writing results to disk:

```python

> evaluation.run(model, output_folder=None)

Reviewer 2: I would always save the results (`v2` will automatically save it to a cache directory, which you of course can specify)

Author: Yes @KennethEnevoldsen  did it changes done now

```

To extract scores programmatically:

```python
results = evaluation.run(model, output_folder=None)

from mteb import MTEBResults
summary = MTEBResults(results).main_scores()

> print(summary)

Reviewer 2: this kinda hides what happens - I would just extract the main score pr task and average them

```

To export all results as a Markdown table:

```python
df = MTEBResults(results).to_markdown()
print(df)
```

---

**Important**: MTEB is for *post-training* benchmarking only.

* Using it during training risks **overfitting** to public benchmarks.

> * It writes to disk and caches aggressively.

Reviewer 2: Is this not ideal?

Author: yes @KennethEnevoldsen  from my understanding and based on discussion we had at the start
This looks ideal for me.

> * Official guidance recommends using SentenceTransformer's built-in evaluators like:

Reviewer 2: If you want to evaluate during training?

Author: @KennethEnevoldsen ok so here the thought was to just show we do evaluation at testing
Although i also thought about it but then idea was to keep things simple and if this works
I would open another issue and then get it done.

Reviewer 2: Hmm, I am not quite sure what you mean, but from my reading, it sounds like you do not recommend MTEB for evaluation (even after training), which I don't think is the intention?

Author: Thanks for the clarification, @KennethEnevoldsen !

You're absolutely right — my intention was not to discourage the use of MTEB for evaluation after training. I’ve now updated the wording to clarify that MTEB is **recommended for post-training evaluation**, but **not ideal during training loops** due to the risk of overfitting and aggressive caching.

Let me know if the revised phrasing works better — happy to tweak further! 🙌

  * `EmbeddingSimilarityEvaluator`
  * `BinaryClassificationEvaluator`

---

## Submitting to the Leaderboard

You can compare your results on the [official leaderboard](https://huggingface.co/spaces/mteb/leaderboard).

Export your scores:

```python
from mteb import MTEBResults
df = MTEBResults(results).to_markdown()

> print(df)

Reviewer 2: Maybe it is better to use the "to_dataframe" method instead?

Author: ok @KennethEnevoldsen  I have moved it to "to_dataframe" method

```

> Follow submission instructions in the [MTEB repo](https://github.com/embeddings-benchmark/mteb).

Reviewer 2: ```suggestion
To add your results to the MTEB Leaderboard, follow the submission instructions in the [MTEB repository](https://github.com/embeddings-benchmark/mteb).
```
